bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+package org.mozartoz.bootcompiler
+package transform
+
+import ast._
+import oz._
+import symtab._
+import scala.collection._
+import com.sun.javafx.fxml.expression.BinaryExpression
+
+sealed class VarState(val onStack: Boolean, val locked: Boolean) {
+}
+object Declared      extends VarState(false, false)
+object Captured      extends VarState(false, false)
+object BoundFirst    extends VarState(true, true)
+object ToInstantiate extends VarState(false, true)
+
+case class VarAndState(v: Variable, state: VarState)
+
+/**
+ * Looks for declarations and the usages of those declarations.
+ * The Transform methods detect local statements and add declared variables to the environment
+ * The Walk methods are collecting their usages and tagging them
+ * 
+ * If a variable is bound after its declaration, we can shortcut it
+ * 
+ * When a proc is declared, it is difficult to track where it will end up. So if a call occurs,
+ * by precaution, all captured variable that were not bound should be instantiated
```

Comment:
It would be interesting to be more precise here: variables not in arguments and not captured by a `proc` in the current scope before the call cannot be accessed by the call (also only if the `proc` is used for the call it can access the captured variables). Would this be difficult or a minor extension?